### ISSUE TITLE
Array-bounds-less Idx wrapper struct?

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#83886]
This is the current version of my Idx wrapper. Surely this code can be further improved. Its purpose is similar to crates like https://crates.io/crates/index-ext , but its usage is cleaner. It just encodes that an index is in-bounds for an array. Each index gets tested when it is created. Then it's usable without bound tests for any array of the same length or longer. This is useful because LLVM is sometimes able to remove those tests at creation time, or it has to perform a test only once (and later the index is usable many times), or even much less than average once, as in the ArraySpan trait below, where you only need to test the two endpoints of a range, and then it creates many Idx without tests at all. This is especially handy when you have a data structure that needs to store array indexes. Later they can be used lots of times without any array bound tests. I have had real performance increase using this Idx (an example with the index-ext crate that can be translated to use Idx: https://github.com/HeroicKatora/index-ext/tree/master/examples ).

A much better (and much more useful) alternative solution is to bite the bullet and add to Rust built-in ranged integers (as in Ada language). If they are well implemented and well integrated with the rest of the language (as they should be), then they can be used as bound-test-free on arrays too, as Idx.

```rust
#![feature(
    const_evaluatable_checked,
    const_fn,
    const_generics,
    const_panic,
    const_trait_impl,
    core_intrinsics,
    step_trait,
    step_trait_ext,
)]
#![allow(incomplete_features)]

use std::{
    intrinsics::assume;
    mem::size_of,
    ops::{Index, IndexMut, RangeInclusive},
};

//---------------------

macro_rules! const_assert {
    ($x:expr $(,)?) => {
        const _: [(); 0 - !{ const ASSERT: bool = $x; ASSERT } as usize] = [];
    };
}

//---------------------

pub trait ToFromUsize {
    fn to_usize(self) -> usize;
    fn from_usize(x: usize) -> Self;
    const MAX: Self;
}
impl const ToFromUsize for usize {
    fn to_usize(self) -> usize { self }
    fn from_usize(x: usize) -> Self { x }
    const MAX: Self = Self::MAX;
}
impl const ToFromUsize for u32 {
    fn to_usize(self) -> usize {
        const_assert!(size_of::<usize>() <= 8);
        debug_assert!(self as u64 <= usize::MAX as u64);
        self as usize
    }
    fn from_usize(x: usize) -> Self {
        const_assert!(size_of::<usize>() <= 8);
        debug_assert!(x as u64 <= Self::MAX as u64);
        x as Self
    }
    const MAX: Self = Self::MAX;
}

//---------------------

pub const fn assert_nonzero(n: usize) -> usize {
    assert!(n > 0);
    n
}

pub const fn is_contained(n: usize, m: usize) -> usize {
    assert!(n <= m);
    n
}

pub const fn is_representable<Ti: ToFromUsize + Copy>(n: usize) -> usize {
    assert!(n <= Ti::MAX.to_usize());
    n
}

mod idx {
    use std::iter::Step;
    use super::{assert_nonzero, assume, ToFromUsize, is_representable};

    #[derive(Debug, Copy, Clone, PartialOrd, PartialEq, Default)]
    #[repr(transparent)]
    pub struct Idx<Ti: ToFromUsize + Copy, const N: usize>(Ti)
    where [(); assert_nonzero(N)]:,
          [(); is_representable::<Ti>(N)]:;

    // TODO Idx::new can't be const, see Issue #79898.
    #[inline(always)]
    pub(crate) const fn idx_new<Ti: ToFromUsize + Copy, const N: usize>(i: Ti) -> Option<Idx<Ti, N>>
    where [(); assert_nonzero(N)]:,
          [(); is_representable::<Ti>(N)]: {
        if i.to_usize() < N { // Issue #79898
            Some(Idx(i))
        } else {
            None
        }
    }

    impl<Ti: ToFromUsize + Copy, const N: usize> Idx<Ti, N>
    where [(); assert_nonzero(N)]:,
          [(); is_representable::<Ti>(N)]: {
        #[inline(always)]
        pub /*const*/ fn new(i: Ti) -> Option<Idx<Ti, N>> { // Issue #79898.
            if i.to_usize() < N {
                Some(Idx(i))
            } else {
                None
            }
        }

        #[inline(always)]
        pub fn from_usize(i: usize) -> Option<Idx<Ti, N>> {
            if i < N {
                Some(Idx(Ti::from_usize(i)))
            } else {
                None
            }
        }

        #[inline(always)]
        pub const unsafe fn new_unchecked(i: Ti) -> Idx<Ti, N> {
            Idx(i)
        }

        #[inline(always)]
        pub fn get(&self) -> Ti {
            unsafe { assume(self.0.to_usize() < N); } // Issue #79898
            self.0
        }
    }

    unsafe impl<Ti: ToFromUsize + Copy, const N: usize> Step for Idx<Ti, N>
    where [(); assert_nonzero(N)]: ,
          [(); is_representable::<Ti>(N)]: ,
          Ti: PartialOrd + Clone {
        fn steps_between(start: &Self, end: &Self) -> Option<usize> {
            Some(end.0.to_usize() - start.0.to_usize())
        }

        fn forward_checked(start: Self, count: usize) -> Option<Self> {
            Self::new(Ti::from_usize(start.0.to_usize() + count))
        }

        fn backward_checked(start: Self, count: usize) -> Option<Self> {
            Some(Self(Ti::from_usize(start.0.to_usize().checked_sub(count)?)))
        }
    }
}

use idx::{Idx, idx_new};

impl<Ti: ToFromUsize + Copy, T, const N: usize, const M: usize> Index<Idx<Ti, N>> for [T; M]
where [(); assert_nonzero(N)]: ,
      [(); is_contained(N, M)]:,
      [(); is_representable::<Ti>(N)]: {
    type Output = T;

    #[inline(always)]
    fn index(&self, i: Idx<Ti, N>) -> &Self::Output {
        unsafe { self.get_unchecked(i.get().to_usize()) }
    }
}

impl<Ti: ToFromUsize + Copy, T, const N: usize, const M: usize> IndexMut<Idx<Ti, N>> for [T; M]
where [(); assert_nonzero(N)]: ,
      [(); is_contained(N, M)]:,
      [(); is_representable::<Ti>(N)]: {
    #[inline(always)]
    fn index_mut(&mut self, i: Idx<Ti, N>) -> &mut Self::Output {
        unsafe { self.get_unchecked_mut(i.get().to_usize()) }
    }
}

// This is only usize to make type inference simpler at usage point.
trait ArraySpan<const N: usize>
where [(); assert_nonzero(N)]: ,
      [(); is_representable::<usize>(N)]: {
    fn span(&self) -> RangeInclusive<Idx<usize, N>>;
}

impl<T, const N: usize> const ArraySpan<N> for [T; N]
where [(); assert_nonzero(N)]:,
      [(); is_representable::<usize>(N)]: {
    #[inline(always)]
    fn span(&self) -> RangeInclusive<Idx<usize, N>> {
        unsafe {
            Idx::new_unchecked(0) ..= Idx::new_unchecked(N - 1)
        }
    }
}

//---------------------

fn main() {}
```

**Comment by jonas-schievink** [comment]
Hi! Additions or changes to the language need to go through our [RFC process](https://github.com/rust-lang/rfcs). Before an RFC is written, the feature can also be discussed in our [internals forum](https://internals.rust-lang.org/) to find other people interested in it.

Closing in favor of an RFC.
